File: 2015/01/Java/src/main/java/Main.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.NoSuchElementException;
import java.util.stream.IntStream;


public class Main {

    public static void main(String[] args) throws IOException {
        var instructions = Files
                .readString(Path.of("..", "input.txt"));
        System.out.println(solvePart1(instructions));
        System.out.println(solvePart2(instructions));
    }

    public static int solvePart1(String instructions) {
        return instructions
                .chars()
                .reduce(0, (f, p) -> f + (p == '(' ? 1 : -1));
    }

    public static int solvePart2(String instructions){
        int floor = 0;
        for(int i = 0; i <= instructions.length(); i++){
            if(instructions.charAt(i) == '(') floor++;
            else if (--floor < 0) return i + 1;
        }
        throw new NoSuchElementException("never reached basement");
    }

    public static int solvePart2_fun(String instructions) {
        // with stream, just for fun
        String enhanced = instructions + ")";
        return IntStream.range(0, enhanced.length())
                       .reduce(0, (f, i) -> f >= 0
                               ? enhanced.charAt(i) == '('
                                       ? f + 1
                                       : f - 1
                               : f == -1
                                       ? -(i * 1_000)
                                       : f)
               / -1_000;
    }
}

File: 2015/01/go/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	input := readInput()
	fmt.Println(solvePart1(input))
	fmt.Println(solvePart2(input))
}

func readInput() string {
	input_raw, err := os.ReadFile("../input.txt")
	if err != nil {
		fmt.Print(err)
	}
	return string(input_raw)
}

func solvePart1(input string) int {
	return strings.Count(input, "(") - strings.Count(input, ")")
}

func solvePart2(input string) int {
	floor := 0
	for n, c := range input {
		if string(c) == "(" {
			floor += 1
		} else {
			floor -= 1
			if floor == -1 {
				return n + 1
			}
		}
	}
	return 0
}
